lib/clickhousex/protocol.ex

```diff
@@ -74,84 +68,88 @@ defmodule Clickhousex.Protocol do
     end
   end
 
-  @doc false
-  @spec reconnect(new_opts :: Keyword.t(), state) :: {:ok, state}
-  def reconnect(new_opts, state) do
-    with :ok <- disconnect("Reconnecting", state),
-         do: connect(new_opts)
+  # TODO find a cleaner way
+  @spec ensure_connected(state) :: {:ok, state} | {:error, Exception.t()}
+  defp ensure_connected(state) do
+    %{conn: conn, conn_opts: conn_opts} = state
+
+    if Client.open?(conn) do
+      {:ok, state}
+    else
+      case connect(conn_opts) do
+        {:ok, _state} = ok -> ok
+        {:error, _conn, reason} -> {:error, reason}
+      end
+    end
   end
 
-  @doc false
-  @spec checkin(state) :: {:ok, state}
+  @impl true
   def checkin(state) do
-    {:ok, state}
+    case ensure_connected(state) do
+      {:ok, _state} = ok -> ok
+      {:error, reason} -> {:disconnect, reason, state}
```

Comment:
But if we can't reconnect in one attempt, an error is logged, and the retries are delegated to dbconnection.